File: .claude/hooks/inject-context.py
#!/usr/bin/env python3
"""inject-context.py — Relevance-Scored Thread Injection (V2)

Hook: UserPromptSubmit (Claude Code) / userPromptSubmitted (Copilot CLI)
Scores threads by file overlap, branch, recency, stuck status, same dev.
Injects top 3-5 most relevant threads under a 300-token budget.
"""

import json
import os
import subprocess
import sys
from datetime import datetime, timezone


def _safe_exit():
    """Output valid JSON and exit cleanly — used when imports fail."""
    try:
        print(json.dumps({"output": ""}))
    except Exception:
        print('{"output": ""}')
    sys.exit(0)


try:
    from sticky_utils import (
        get_memory_path, load_json, save_json, get_user, get_branch,
        get_resume_thread_id, find_thread_by_id, get_session_id,
        append_audit_line,
    )
except Exception:
    if __name__ == "__main__":
        _safe_exit()


def get_recently_modified_files():
    """Get files modified in recent commits (proxy for current work area)."""
    files = set()
    try:
        result = subprocess.run(
            ["git", "diff", "--name-only", "HEAD~5"],
            capture_output=True, text=True, timeout=5
        )
        if result.returncode == 0:
            files.update(f.strip() for f in result.stdout.strip().split("\n") if f.strip())
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
        pass
    # Also include uncommitted changes
    try:
        result = subprocess.run(
            ["git", "diff", "--name-only"],
            capture_output=True, text=True, timeout=5
        )
        if result.returncode == 0:
            files.update(f.strip() for f in result.stdout.strip().split("\n") if f.strip())
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
        pass
    return files


# ── Keyword extraction ─────────────────────────────────────

def extract_keywords(prompt):
    keywords = set()
    words = prompt.lower().replace("/", " ").replace("\\", " ").replace(".", " ").split()

    for word in words:
        cleaned = word.strip("()[]{}\"'`,;:")
        if len(cleaned) >= 2:
            keywords.add(cleaned)

    for token in prompt.split():
        token = token.strip("()[]{}\"'`,;:")
        if "/" in token or "\\" in token or "." in token:
            keywords.add(token.lower())
            parts = token.replace("\\", "/").split("/")
            for part in parts:
                if len(part) >= 2:
                    keywords.add(part.lower())

    return keywords


# ── Relevance scoring ──────────────────────────────────────

def score_thread(thread, recently_modified, current_branch, current_user, prompt_keywords):
    """Score a thread for relevance. Higher = more relevant."""
    if thread.get("status") in ("expired", "stale"):
        return -1

    score = 0.0
    now = datetime.now(timezone.utc)

    # File overlap (weight 3) — highest signal
    thread_files = set(thread.get("files_touched", []))
    overlap = thread_files & recently_modified
    if overlap:
        score += 3 * min(len(overlap), 5)

    # Prompt keyword match against thread files
    for file_path in thread_files:
        path_lower = file_path.lower()
        path_parts = set(path_lower.replace("\\", "/").replace(".", "/").split("/"))
        for kw in prompt_keywords:
            if kw in path_lower or kw in path_parts:
                score += 1
                break

    # Branch match (weight 2)
    if current_branch and thread.get("branch") == current_branch:
        score += 2

    # Recency (weight 2) — decay over days
    ts_field = (
        thread.get("last_activity_at")
        or thread.get("updated_at")
        or thread.get("created_at", "")
    )
    if ts_field:
        try:
            ts = datetime.fromisoformat(ts_field.replace("Z", "+00:00"))
            days_ago = max((now - ts).total_seconds() / 86400, 0)
            recency = max(2 - (days_ago * 0.2), 0)
            score += recency
        except (ValueError, TypeError):
            pass

    # Stuck boost (+2)
    if thread.get("status") == "stuck":
        score += 2

    # Same developer (weight 1)
    thread_user = thread.get("user") or thread.get("author", "")
    if thread_user == current_user:
        score += 1

    return score


# ── Formatting ─────────────────────────────────────────────

def format_top_thread(thread):
    """Full detail for top-scoring thread."""
    user = thread.get("user") or thread.get("author", "unknown")
    status = thread.get("status", "open")
    files = ", ".join(thread.get("files_touched", [])[:3])
    narrative = thread.get("narrative", "")
    failed = thread.get("failed_approaches", [])
    branch = thread.get("branch", "")

    status_icon = "🔴" if status == "stuck" else "🟢" if status == "open" else "⚪"
    line = f"{status_icon} {files} · {user}"
    if branch:
        line += f" · {branch}"
    ts_field = thread.get("last_activity_at") or thread.get("updated_at", "")
    if ts_field:
        line += f" · {_relative_time(ts_field)}"
    line += "\n"

    if narrative:
        line += narrative[:200] + "\n"
    elif thread.get("last_note"):
        line += thread["last_note"][:150] + "\n"

    # Include conversation prompts for cross-tool context
    prompts = thread.get("prompts", [])
    if prompts:
        line += f"Conversation ({len(prompts)} prompt(s)):\n"
        for i, p in enumerate(prompts[:5], 1):
            line += f"  {i}. {p[:120]}\n"
        if len(prompts) > 5:
            line += f"  ... and {len(prompts) - 5} more\n"

    if failed:
        line += f"{len(failed)} failed approach(es)."
        line += f" Full context: ask get_session_context(\"{thread.get('id', '')}\")\n"

    return line.strip()


def format_summary_thread(thread):
    """One-liner for matches 2-5."""
    user = thread.get("user") or thread.get("author", "unknown")
    status = thread.get("status", "open")
    files = ", ".join(thread.get("files_touched", [])[:2])
    note = thread.get("last_note", "")[:60]

    status_icon = "🔴" if status == "stuck" else "🟢" if status == "open" else "⚪"
    line = f"{status_icon} {files} · {user}"
    if note:
        line += f" — {note}"
    return line


def _relative_time(ts_str):
    try:
        ts = datetime.fromisoformat(ts_str.replace("Z", "+00:00"))
        now = datetime.now(timezone.utc)
        delta = now - ts
        hours = delta.total_seconds() / 3600
        if hours < 1:
            return f"{int(delta.total_seconds() / 60)}min ago"
        elif hours < 24:
            return f"{int(hours)}hrs ago"
        else:
            return f"{int(hours / 24)}d ago"
    except (ValueError, TypeError):
        return ""


# ── Main ───────────────────────────────────────────────────

def main():
    try:
        hook_input = json.loads(sys.stdin.read()) if not sys.stdin.isatty() else {}
    except (json.JSONDecodeError, Exception):
        hook_input = {}

    prompt = hook_input.get("prompt", hook_input.get("user_prompt", ""))
    if not prompt:
        print(json.dumps({"output": ""}))
        return

    # Log user prompt to audit trail so session-end can use it
    session_id = get_session_id(hook_input)
    try:
        append_audit_line({
            "type": "user_prompt",
            "user": get_user(),
            "ts": datetime.now(timezone.utc).isoformat(),
            "session_id": session_id,
            "prompt": prompt[:500],
        })
    except (OSError, IOError):
        pass  # Audit write failure must not block context injection

    memory = load_json(get_memory_path(), {"version": "2", "threads": []})
    threads = memory.get("threads", [])

    # Only consider live threads
    live = [t for t in threads if t.get("status") in ("open", "stuck", "closed")]
    if not live:
        print(json.dumps({"output": ""}))
        return

    keywords = extract_keywords(prompt)
    current_branch = get_branch()
    current_user = get_user()
    recently_modified = get_recently_modified_files()

    # Score all live threads
    resume_thread_id = get_resume_thread_id()
    memory_dirty = False

    # Mid-session resume: reopen the thread if signal file exists
    if resume_thread_id:
        resumed = find_thread_by_id(threads, resume_thread_id)
        if resumed and resumed.get("status") != "open":
            resumed["status"] = "open"
            resumed["last_activity_at"] = datetime.now(timezone.utc).isoformat()
            session_id = get_session_id(hook_input)
            related = resumed.setdefault("related_session_ids", [])
            if session_id not in related:
                related.append(session_id)
            memory_dirty = True

    scored = []
    for t in live:
        s = score_thread(t, recently_modified, current_branch, current_user, keywords)
        # Resumed thread always surfaces first
        if resume_thread_id and t.get("id") == resume_thread_id:
            s = max(s, 0) + 10
        if s > 0:
            scored.append((s, t))

    if memory_dirty:
        try:
            save_json(get_memory_path(), memory)
        except (OSError, IOError):
            pass  # Memory save failure must not block context injection

    scored.sort(key=lambda x: x[0], reverse=True)

    if not scored:
        print(json.dumps({"output": ""}))
        return

    # Token budget: ~300 tokens max
    MAX_TOKENS = 300
    token_count = 10  # header overhead

    output_lines = []
    threads_shown = 0

    for i, (score, thread) in enumerate(scored[:5]):
        if i == 0:
            block = format_top_thread(thread)
        else:
            block = format_summary_thread(thread)

        token_count += len(block) // 4
        if token_count > MAX_TOKENS:
            remaining = len(scored) - i
            if remaining > 0:
                output_lines.append(f"... and {remaining} more relevant threads")
            break
        output_lines.append(block)
        threads_shown += 1

    header = f"[STICKY NOTE — {threads_shown} relevant thread{'s' if threads_shown != 1 else ''}]\n"
    output_lines.insert(0, header)

    output = "\n".join(output_lines).strip()
    print(json.dumps({"output": output}, ensure_ascii=False))


if __name__ == "__main__":
    try:
        main()
    except BaseException:
        _safe_exit()

File: templates/hooks/inject-context.js
#!/usr/bin/env node
"use strict";
/**
 * inject-context.js — Relevance-Scored Thread Injection (V2)
 *
 * Hook: UserPromptSubmit (Claude Code) / userPromptSubmitted (Copilot CLI)
 * Scores threads by file overlap, branch, recency, stuck status, same dev.
 * Injects top 3-5 most relevant threads under a configurable token budget.
 */

const { execSync } = require("child_process");

// ── Output helpers ────────────────────────────────────────

function _isCopilotCli() {
  return process.argv.includes("--copilot-cli") || !!process.env.COPILOT_CLI;
}

function _emit(text) {
  if (text === undefined) text = "";
  if (_isCopilotCli()) {
    process.stdout.write(JSON.stringify({ additionalContext: text }) + "\n");
  } else {
    process.stdout.write(JSON.stringify({ output: text }) + "\n");
  }
}

function _safeExit() {
  try {
    _emit("");
  } catch (_) {
    process.stdout.write('{"output": ""}\n');
  }
  process.exit(0);
}

// ── Import sticky-utils ──────────────────────────────────

let utils;
try {
  utils = require("./sticky-utils.js");
} catch (_) {
  _safeExit();
}
const {
  getMemoryPath, loadJson, saveJson, getUser, getBranch,
  getResumeThreadId, findThreadById, getSessionId,
  appendAuditLine, detectTool, getConfigPath,
  isThreadInjected, markThreadInjected, normalizeSep,
} = utils;

// ── Git helpers ───────────────────────────────────────────

function getRecentlyModifiedFiles() {
  const files = new Set();
  // Try HEAD~5 first; fall back to HEAD~1 for shallow repos
  const diffTargets = ["HEAD~5", "HEAD~1"];
  for (const target of diffTargets) {
    try {
      const result = execSync(`git diff --name-only ${target}`, {
        encoding: "utf-8",
        timeout: 5000,
        stdio: ["pipe", "pipe", "pipe"],
      });
      for (const f of result.trim().split(/\r?\n/)) {
        const trimmed = f.trim();
        if (trimmed) files.add(trimmed);
      }
      break;
    } catch (_) {
      // target doesn't exist, try next
    }
  }
  try {
    const result = execSync("git diff --name-only", {
      encoding: "utf-8",
      timeout: 5000,
      stdio: ["pipe", "pipe", "pipe"],
    });
    for (const f of result.trim().split(/\r?\n/)) {
      const trimmed = f.trim();
      if (trimmed) files.add(trimmed);
    }
  } catch (_) {
    // ignore
  }
  return files;
}

// ── Keyword extraction ────────────────────────────────────

function extractKeywords(prompt) {
  const keywords = new Set();
  const words = prompt.toLowerCase().replace(/[/\\.]/g, " ").split(/\s+/);
  for (const word of words) {
    const cleaned = word.replace(/[()[\]{}"'`,;:]/g, "");
    if (cleaned.length >= 2) {
      keywords.add(cleaned);
    }
  }
  for (let token of prompt.split(/\s+/)) {
    token = token.replace(/^[()[\]{}"'`,;:]+/, "").replace(/[()[\]{}"'`,;:]+$/, "");
    if (token.includes("/") || token.includes("\\") || token.includes(".")) {
      keywords.add(token.toLowerCase());
      const parts = normalizeSep(token).split("/");
      for (const part of parts) {
        if (part.length >= 2) {
          keywords.add(part.toLowerCase());
        }
      }
    }
  }
  return keywords;
}

// ── Scoring ───────────────────────────────────────────────

function scoreThread(thread, recentlyModified, currentBranch, currentUser, promptKeywords) {
  if (thread.status === "expired" || thread.status === "stale") {
    return -1;
  }

  let score = 0.0;
  const now = new Date();

  // File overlap (weight 3)
  const threadFiles = new Set(thread.files_touched || []);
  let overlapCount = 0;
  for (const f of threadFiles) {
    if (recentlyModified.has(f)) overlapCount++;
  }
  if (overlapCount > 0) {
    score += 3 * Math.min(overlapCount, 5);
  }

  // Prompt keyword match against thread files
  for (const filePath of threadFiles) {
    const pathLower = filePath.toLowerCase();
    const pathParts = new Set(normalizeSep(pathLower).replace(/\./g, "/").split("/"));
    for (const kw of promptKeywords) {
      if (pathLower.includes(kw) || pathParts.has(kw)) {
        score += 1;
        break;
      }
    }
  }

  // Branch match (weight 2)
  if (currentBranch && thread.branch === currentBranch) {
    score += 2;
  }

  // Recency (weight 2) — decay over days
  const tsField = thread.last_activity_at || thread.updated_at || thread.created_at || "";
  if (tsField) {
    try {
      const ts = new Date(tsField.replace("Z", "+00:00"));
      if (!isNaN(ts.getTime())) {
        const daysAgo = Math.max((now - ts) / 86400000, 0);
        const recency = Math.max(2 - daysAgo * 0.2, 0);
        score += recency;
      }
    } catch (_) {
      // ignore
    }
  }

  // Stuck boost (+2)
  if (thread.status === "stuck") {
    score += 2;
  }

  // Same developer (weight 1)
  const threadUser = thread.user || thread.author || "";
  if (threadUser === currentUser) {
    score += 1;
  }

  return score;
}

// ── Formatting ────────────────────────────────────────────

function _relativeTime(tsStr) {
  try {
    const ts = new Date(tsStr);
    if (isNaN(ts.getTime())) return "";
    const now = new Date();
    const deltaMs = now - ts;
    const hours = deltaMs / 3600000;
    if (hours < 1) {
      return Math.floor(deltaMs / 60000) + "min ago";
    } else if (hours < 24) {
      return Math.floor(hours) + "hrs ago";
    } else {
      return Math.floor(hours / 24) + "d ago";
    }
  } catch (_) {
    return "";
  }
}

function formatThread(thread, detailed) {
  const user = thread.user || thread.author || "unknown";
  const status = thread.status || "open";
  const statusTag = status === "stuck" ? "[STUCK]" : status === "open" ? "[OPEN]" : "[CLOSED]";
  const fileLimit = detailed ? 3 : 2;
  const files = (thread.files_touched || []).slice(0, fileLimit).join(", ");

  let line = statusTag + " " + files + " . " + user;

  if (!detailed) {
    const note = (thread.last_note || "").substring(0, 60);
    if (note) line += " -- " + note;
    return line;
  }

  const branch = thread.branch || "";
  if (branch) line += " . " + branch;
  const tsField = thread.last_activity_at || thread.updated_at || "";
  if (tsField) line += " . " + _relativeTime(tsField);
  line += "\n";

  if (thread.narrative) {
    line += thread.narrative.substring(0, 200) + "\n";
  } else if (thread.last_note) {
    line += thread.last_note.substring(0, 150) + "\n";
  }

  const prompts = thread.prompts || [];
  if (prompts.length > 0) {
    line += "Conversation (" + prompts.length + " prompt(s)):\n";
    for (let i = 0; i < Math.min(prompts.length, 5); i++) {
      line += "  " + (i + 1) + ". " + prompts[i].substring(0, 120) + "\n";
    }
    if (prompts.length > 5) line += "  ... and " + (prompts.length - 5) + " more\n";
  }

  const failed = thread.failed_approaches || [];
  if (failed.length > 0) {
    line += failed.length + " failed approach(es).";
    line += ' Full context: ask get_session_context("' + (thread.id || "") + '")\n';
  }

  return line.trim();
}

// ── Main ──────────────────────────────────────────────────

function main() {
  let hookInput = {};
  try {
    const raw = require("fs").readFileSync(0, "utf-8");
    if (raw.trim()) {
      hookInput = JSON.parse(raw);
    }
  } catch (_) {
    hookInput = {};
  }

  const prompt = hookInput.prompt || hookInput.user_prompt || "";
  if (!prompt) {
    _emit("");
    return;
  }

  const sessionId = getSessionId(hookInput);

  function _auditInject(result, threadsScored, threadsInjected, topScores, error) {
    threadsScored = threadsScored || 0;
    threadsInjected = threadsInjected || 0;
    try {
      const entry = {
        type: "inject_result",
        user: getUser(),
        ts: new Date().toISOString(),
        session_id: sessionId,
        prompt: prompt.substring(0, 120),
        result: result,
        threads_scored: threadsScored,
        threads_injected: threadsInjected,
      };
      if (topScores) entry.top_scores = topScores;
      if (error) entry.error = String(error).substring(0, 200);
      appendAuditLine(entry);
    } catch (_) {
      // ignore
    }
  }

  // Audit the user prompt
  try {
    appendAuditLine({
      type: "user_prompt",
      user: getUser(),
      ts: new Date().toISOString(),
      session_id: sessionId,
      prompt: prompt.substring(0, 500),
    });
  } catch (_) {
    // ignore
  }

  const memory = loadJson(getMemoryPath(), { version: "2", threads: [] });
  const threads = memory.threads || [];

  const live = threads.filter(
    (t) => t.status === "open" || t.status === "stuck" || t.status === "closed"
  );
  if (live.length === 0) {
    _auditInject("no_live_threads");
    _emit("");
    return;
  }

  const keywords = extractKeywords(prompt);
  const currentBranch = getBranch();
  const currentUser = getUser();
  const recentlyModified = getRecentlyModifiedFiles();

  const resumeThreadId = getResumeThreadId();
  let memoryDirty = false;

  if (resumeThreadId) {
    const resumed = findThreadById(threads, resumeThreadId);
    if (resumed && resumed.status !== "open") {
      resumed.status = "open";
      resumed.last_activity_at = new Date().toISOString();
      const related = resumed.related_session_ids || [];
      if (!related.includes(sessionId)) {
        related.push(sessionId);
      }
      resumed.related_session_ids = related;
      const chain = resumed.resume_chain || [];
      const alreadyInChain = chain.some((e) => e.session_id === sessionId);
      if (!alreadyInChain) {
        const aiTool = detectTool(hookInput);
        const prevSession =
          chain.length > 0
            ? chain[chain.length - 1].session_id
            : resumed.session_id || "";
        chain.push({
          session_id: sessionId,
          tool: aiTool,
          started_at: new Date().toISOString(),
          ended_at: null,
          resumed_from: prevSession,
        });
      }
      resumed.resume_chain = chain;
      memoryDirty = true;
    }
  }

  // Score threads
  const scored = [];
  for (const t of live) {
    // V2.5: Skip threads already injected this session by PreToolUse or session-start
    if (isThreadInjected(t.id)) continue;

    let s = scoreThread(t, recentlyModified, currentBranch, currentUser, keywords);
    if (resumeThreadId && t.id === resumeThreadId) {
      s = Math.max(s, 0) + 10;
    }
    if (s > 0) {
      scored.push([s, t]);
    }
  }

  if (memoryDirty) {
    try {
      saveJson(getMemoryPath(), memory);
    } catch (_) {
      // ignore
    }
  }

  scored.sort((a, b) => b[0] - a[0]);

  // Debug / scoring block
  const debugLines = [
    "\n---\n[STICKY-NOTE] [scoring] prompt: " + prompt.substring(0, 80),
  ];
  debugLines.push(
    "  keywords: " +
      Array.from(keywords)
        .sort()
        .slice(0, 15)
        .join(", ")
  );
  debugLines.push("  branch: " + currentBranch + "  user: " + currentUser);
  debugLines.push(
    "  recently_modified: " +
      Array.from(recentlyModified)
        .sort()
        .slice(0, 10)
        .join(", ")
  );
  debugLines.push(
    "  live threads: " + live.length + "  scored>0: " + scored.length
  );
  for (const [s, t] of scored.slice(0, 10)) {
    const tid = (t.id || "").substring(0, 8);
    const st = t.status || "?";
    const files = (t.files_touched || []).slice(0, 3);
    const tool = t.tool || "?";
    debugLines.push(
      "  " +
        s.toFixed(1).padStart(5) +
        "  " +
        tid +
        "  [" +
        st +
        "] " +
        tool +
        "  files=" +
        JSON.stringify(files)
    );
  }
  const scoringBlock = debugLines.join("\n");

  if (scored.length === 0) {
    _auditInject("no_scored_threads", 0);
    _emit("");
    return;
  }

  const config = loadJson(getConfigPath(), {});
  const MAX_TOKENS = config.inject_token_budget || 1000;
  const _estimateTokens = (str) => Math.floor(str.length / 4);

  // Reserve budget for the scoring debug block and header overhead
  const scoringTokens = _estimateTokens(scoringBlock);
  const HEADER_RESERVE = 15; // "[STICKY NOTE -- N relevant thread(s)]"
  let tokenCount = HEADER_RESERVE + scoringTokens;

  const outputLines = [];
  let threadsShown = 0;

  for (let i = 0; i < Math.min(scored.length, 5); i++) {
    const [score, thread] = scored[i];
    const block =
      i === 0 ? formatThread(thread, true) : formatThread(thread, false);
    const blockTokens = _estimateTokens(block);
    if (tokenCount + blockTokens > MAX_TOKENS) {
      const remaining = scored.length - i;
      if (remaining > 0) {
        const overflowMsg = "... and " + remaining + " more relevant threads";
        tokenCount += _estimateTokens(overflowMsg);
        outputLines.push(overflowMsg);
      }
      break;
    }
    tokenCount += blockTokens;
    outputLines.push(block);
    threadsShown++;

    // V2.5: Mark as injected so PreToolUse won't re-inject
    try { markThreadInjected(thread.id, sessionId); } catch (_) {}
  }

  const header =
    "[STICKY-NOTE] " +
    threadsShown +
    " relevant thread" +
    (threadsShown !== 1 ? "s" : "") +
    " injected\n";
  outputLines.unshift(header);

  let output = outputLines.join("\n").trim();
  output += scoringBlock;

  _auditInject(
    "injected",
    scored.length,
    threadsShown,
    scored.slice(0, 5).map(([s, t]) => ({
      id: (t.id || "").substring(0, 8),
      score: Math.round(s * 10) / 10,
    }))
  );
  _emit(output);
}

// ── Entry point ───────────────────────────────────────────

if (require.main === module) {
  try {
    main();
  } catch (exc) {
    try {
      appendAuditLine({
        type: "inject_result",
        ts: new Date().toISOString(),
        result: "error",
        error: String(exc).substring(0, 200),
      });
    } catch (_) {
      // ignore
    }
    _safeExit();
  }
}
